src/network/invthread.py

```diff
@@ -12,6 +12,27 @@
 import protocol
 import state
 
+
+def handleExpiredDandelion(expired):
+    """For expired dandelion objects, mark all remotes as not having
+       the object"""
+    if not expired:
+        return
+    for i in \
+        BMConnectionPool().inboundConnections.values() + \
+            BMConnectionPool().outboundConnections.values():
+        if not i.fullyEstablished:
+            continue
+        for x in expired:
+            streamNumber, hashid, _ = x
+            try:
+                del i.objectsNewToMe[hashid]
+            except KeyError:
+                if streamNumber in i.streams:
+                    with i.objectsNewToThemLock:
+                        i.objectsNewToThem[hashid] = time()
```

Comment:
Maybe this function should be a method of `InvThread`, like `handleLocallyGenerated` below?

Unfortunately I don't understand the meaning of `i.objectsNewToThem[hashid] = time()` ):

Reply:
If I make it into a method, pylint complains that "method could be a function".

The time in objectsNewToThem is for expiration, so that if the peer doesn't request the object the tracking doesn't occupy memory forever.

Reply:
Also this seems to duplicate `handleLocallyGenerated` if child is None:

```python
                try:
                    data = invQueue.get(False)
                    chunk.append((data[0], data[1]))
                    # locally generated
                    if len(data) == 2 or data[2] is None:
                        self.handleLocallyGenerated(data[0], data[1])
                except Queue.Empty:
                    break
```

Reply:
No, handleLocallyGenerated advertises to a single child stem, expiration to everyone.

Reply:
> If I make it into a method, pylint complains that "method could be a function".
> 

So maybe there is no need for separate function in `invthread` module? Let's make it part of `Dandelion().expire()`

```diff
diff --git a/src/network/dandelion.py b/src/network/dandelion.py
index 60d84cdf..bb5cdee7 100644
--- a/src/network/dandelion.py
+++ b/src/network/dandelion.py
@@ -130,7 +130,22 @@ class Dandelion():
             for row in toDelete:
                 self.removeHash(row[1], 'expiration')
                 invQueue.put(row)
-        return toDelete
+
+        # For expired dandelion objects, mark all remotes as not having
+        # the object
+        for connection in (
+            network.connectionpool.BMConnectionPool().inboundConnections.values() +
+            network.connectionpool.BMConnectionPool().outboundConnections.values()
+        ):
+            if not connection.fullyEstablished:
+                continue
+            for streamNumber, hashid, _ in toDelete:
+                try:
+                    del connection.objectsNewToMe[hashid]
+                except KeyError:
+                    if streamNumber in connection.streams:
+                        with connection.objectsNewToThemLock:
+                            connection.objectsNewToThem[hashid] = time()
 
     def reRandomiseStems(self):
         with self.lock:
diff --git a/src/network/invthread.py b/src/network/invthread.py
index 6f6f1364..37b22cbf 100644
--- a/src/network/invthread.py
+++ b/src/network/invthread.py
@@ -13,26 +13,6 @@ import protocol
 import state
 
 
-def handleExpiredDandelion(expired):
-    """For expired dandelion objects, mark all remotes as not having
-       the object"""
-    if not expired:
-        return
-    for i in \
-        BMConnectionPool().inboundConnections.values() + \
-            BMConnectionPool().outboundConnections.values():
-        if not i.fullyEstablished:
-            continue
-        for x in expired:
-            streamNumber, hashid, _ = x
-            try:
-                del i.objectsNewToMe[hashid]
-            except KeyError:
-                if streamNumber in i.streams:
-                    with i.objectsNewToThemLock:
-                        i.objectsNewToThem[hashid] = time()
-
-
 class InvThread(threading.Thread, StoppableThread):
     def __init__(self):
         threading.Thread.__init__(self, name="InvBroadcaster")
@@ -53,7 +33,7 @@ class InvThread(threading.Thread, StoppableThread):
             chunk = []
             while True:
                 # Dandelion fluff trigger by expiration
-                handleExpiredDandelion(Dandelion().expire())
+                Dandelion().expire()
                 try:
                     data = invQueue.get(False)
                     chunk.append((data[0], data[1]))
```

Reply:
I'd also suggest to replace this annoying expression
```python
network.connectionpool.BMConnectionPool().inboundConnections.values() +
network.connectionpool.BMConnectionPool().outboundConnections.values()
```

with something like `BMConnectionPool().connections`. Unfortunately [the generator](../../../g1itch/PyBitmessage/commit/2dc3586) I wrote not works because connections got changed while iterating ):

Reply:
> So maybe there is no need for separate function in `invthread` module? Let's make it part of `Dandelion().expire()`

That's what I wanted to do originally but then for some reason decided not to. Now in retrospect, that would probably be better.
> 
> ```diff
> diff --git a/src/network/dandelion.py b/src/network/dandelion.py
> index 60d84cdf..bb5cdee7 100644
> --- a/src/network/dandelion.py
> +++ b/src/network/dandelion.py
> @@ -130,7 +130,22 @@ class Dandelion():
>              for row in toDelete:
>                  self.removeHash(row[1], 'expiration')
>                  invQueue.put(row)
> -        return toDelete
> +
> +        # For expired dandelion objects, mark all remotes as not having
> +        # the object
> +        for connection in (
> +            network.connectionpool.BMConnectionPool().inboundConnections.values() +
> +            network.connectionpool.BMConnectionPool().outboundConnections.values()
> +        ):
> +            if not connection.fullyEstablished:
> +                continue
> +            for streamNumber, hashid, _ in toDelete:
> +                try:
> +                    del connection.objectsNewToMe[hashid]
> +                except KeyError:
> +                    if streamNumber in connection.streams:
> +                        with connection.objectsNewToThemLock:
> +                            connection.objectsNewToThem[hashid] = time()
>  
>      def reRandomiseStems(self):
>          with self.lock:
> diff --git a/src/network/invthread.py b/src/network/invthread.py
> index 6f6f1364..37b22cbf 100644
> --- a/src/network/invthread.py
> +++ b/src/network/invthread.py
> @@ -13,26 +13,6 @@ import protocol
>  import state
>  
>  
> -def handleExpiredDandelion(expired):
> -    """For expired dandelion objects, mark all remotes as not having
> -       the object"""
> -    if not expired:
> -        return
> -    for i in \
> -        BMConnectionPool().inboundConnections.values() + \
> -            BMConnectionPool().outboundConnections.values():
> -        if not i.fullyEstablished:
> -            continue
> -        for x in expired:
> -            streamNumber, hashid, _ = x
> -            try:
> -                del i.objectsNewToMe[hashid]
> -            except KeyError:
> -                if streamNumber in i.streams:
> -                    with i.objectsNewToThemLock:
> -                        i.objectsNewToThem[hashid] = time()
> -
> -
>  class InvThread(threading.Thread, StoppableThread):
>      def __init__(self):
>          threading.Thread.__init__(self, name="InvBroadcaster")
> @@ -53,7 +33,7 @@ class InvThread(threading.Thread, StoppableThread):
>              chunk = []
>              while True:
>                  # Dandelion fluff trigger by expiration
> -                handleExpiredDandelion(Dandelion().expire())
> +                Dandelion().expire()
>                  try:
>                      data = invQueue.get(False)
>                      chunk.append((data[0], data[1]))
> ```

Reply:
> I'd also suggest to replace this annoying expression

Yes, that would be great, except as you wrote yourself, due to multithreading it creates a problem with generators. One option would be to use locking, but this would negatively improve performance (or could even cause a deadlock in some cases). If on the other hand you do a list of references, it seems to work fine. Perhaps then a method that returns a combined list of references so that you don't have to `+` everytime would work.